search/search_quality/assessment_tool/context.cpp

```diff
@@ -8,6 +15,79 @@ void Context::Clear()
   m_initialized = false;
 }
 
+search::Sample Context::MakeSample(search::FeatureLoader & loader) const
+{
+  search::Sample sample = m_sample;
+
+  if (!m_initialized)
+    return sample;
+
+  auto const & relevances = m_edits.GetRelevances();
+
+  auto & results = sample.m_results;
+  results.clear();
+
+  CHECK_EQUAL(m_goldenMatching.size(), m_sample.m_results.size(), ());
+  CHECK_EQUAL(m_actualMatching.size(), relevances.size(), ());
+  CHECK_EQUAL(m_actualMatching.size(), m_results.GetCount(), ());
+
+  // Iterates over original (loaded from the file with search samples)
+  // results first.
+  for (size_t i = 0; i < m_sample.m_results.size(); ++i)
+  {
+    auto const j = m_goldenMatching[i];
+
+    // Some results weren't matched, so they weren't displayed to the
+    // assessor. But we want to keep them.
+    if (j == search::Matcher::kInvalidId)
+    {
+      results.push_back(m_sample.m_results[i]);
+      continue;
+    }
+
+    // No need to keep irrelevant results.
+    if (relevances[j] == search::Sample::Result::Relevance::Irrelevant)
+      continue;
+
+    auto result = m_sample.m_results[i];
+    result.m_relevance = relevances[j];
+    results.push_back(result);
+  }
+
+  // Iterates over results retrieved during assessment.
+  for (size_t i = 0; i < m_results.GetCount(); ++i)
+  {
+    auto const j = m_actualMatching[i];
+    if (j != search::Matcher::kInvalidId)
+    {
+      // This result was processed by the loop above.
+      continue;
+    }
+
+    // No need to keep irrelevant results.
+    if (relevances[i] == search::Sample::Result::Relevance::Irrelevant)
+      continue;
+
+    auto const & result = m_results.GetResult(i);
+    // No need in non-feature results.
+    if (result.GetResultType() != search::Result::RESULT_FEATURE)
+      continue;
+
+    FeatureType ft;
+    CHECK(loader.Load(result.GetFeatureID(), ft), ());
+    results.push_back(search::Sample::Result::Build(ft, relevances[i]));
```

Comment:
Not a big performance loss. Move ctor will be used here, moreover, Build static method looks much more readable for me, as a factory method.